### PR TITLE
[#12011] Instructor Viewing Student Activities: Bug Fix and Enhancement

### LINKED ISSUE #12011: Instructor viewing student activities: distinguish between view and submit
Context: the unreleased 'view student activity feature' mentioned in https://github.com/TEAMMATES/teammates/pull/11845

image

Suggestion: even if the `session access and submission` option is chosen, the results should differentiate between view access and submission. Showing the activity as `viewed OR submitted` is less useful than it can otherwise be.

### PR DESCRIPTION
**Pull request by domlimm**
Fixes #12011.

**Outline of Solution**

- [X] Fix bug with wrong status prefix based off drop down selection
- [X] Show latest log for each selected activity type instead of showing latest log of all types

Previously, we generate the status prefix for each activity log based on the selected activity type from the drop down list in the filter container. This showed the wrong description for the logs.

We should instead use the `FeedbackSessionLogEntry`'s type to generate the status prefix/description for these logs.

Additionally, we only show the latest log (be it access or submission) for when the `session access and submission` activity type has been selected for filtering. The image below shows the output after fixing the bug of wrong status prefix due to using the drop down selection's type to generate the prefix.

image

Now, we will show the latest log for both access and submission feedback session log type.

image

### CODE REVIEW

File: src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts:303]: I think we can use `let statusPrefix = this.logTypeToActivityDisplay(entry.feedbackSessionLogType);` if `studentKey in this.studentToLog`, otherwise we use `this.logTypeToActivityDisplay(this.formModel.logType);` for inactivities. This could get rid of the logic from line 330-395

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]

File: src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts
Comment (domlimm) [reply src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts:303]: hey hieu, thank you for reviewing this PR! lines 330-395 adds the latest log for either access or submission activity type.

this addition is based on what's added in the if clause `studentKey in this.studentToLog`.

if we are initially adding an access activity type, in lines 330-395, we will search through the logs for this student and get the latest submission activity log, and vice versa.

Comment (daongochieu2810) [reply src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts:303]: Can we move this logic to the backend instead? So we don't query for irrelevant logs on the frontend

Comment (domlimm) [reply src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts:303]: i understand that this chunk of codes is lengthy on the frontend. i checked and saw that we query all the logs based on the params we added when calling the logs service. in turn, it returns all the relevant logs. may i ask how should we move this logic to the backend? currently, we pull all logs based on the params we specify.

**here below is the current change in the PR:**

**for context:**
for our case, the edge case/enhancement, we'd want to display both `access` and `submission` logs. however, the initial implementation shows only the latest of either.

i did not want to affect the current implementation as i thought it was a really smart way to accommodate for inactivities too whenever specified in the control panel.

so, i used the logs that were returned from the log service and naively pushed the missing log into `logRowsData`, as this seems to be the only 'edge' case as the rest are well handled.

Comment (daongochieu2810) [reply src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts:303]: @domlimm Let's put this on hold until this is merged first, then it will be much easier for you
https://github.com/TEAMMATES/teammates/pull/11857

Comment (domlimm) [reply src/web/app/pages-instructor/instructor-student-activity-logs/instructor-student-activity-logs.component.ts:303]: noted! thanks hieu! let me look at the PR again in the meantime!

**Comment by wkurniawan07** [comment]
Will be closing this for now, we will revisit the feature in greater detail in a few weeks time

**Pull request closed without merging**
